packages/nextjs/src/utils/instrumentServer.ts

```diff
@@ -182,40 +198,81 @@ function makeWrappedReqHandler(origReqHandler: ReqHandler): WrappedReqHandler {
+      const currentScope = Sentry.getCurrentHub().getScope();
+
+      if (currentScope) {
+        currentScope.addEventProcessor(event => addRequestDataToEvent(event, req));
+
+        // We only want to record page and API requests
+        if (hasTracingEnabled() && shouldTraceRequest(req.url, publicDirFiles)) {
+          // pull off query string, if any
+          const reqPath = req.url.split('?')[0];
+
+          // requests for pages will only ever be GET requests, so don't bother to include the method in the transaction
+          // name; requests to API routes could be GET, POST, PUT, etc, so do include it there
+          const namePrefix = req.url.startsWith('/api') ? `${(req.method || 'GET').toUpperCase()} ` : '';
+
+          const transaction = Sentry.startTransaction({
+            name: `${namePrefix}${reqPath}`,
+            op: 'http.server',
+            metadata: { request: req },
+          });
+
+          currentScope.setSpan(transaction);
+
+          res.once('finish', () => {
+            const transaction = getActiveTransaction();
+            if (transaction) {
               transaction.setHttpStatus(res.statusCode);
-              transaction.finish();
-            });
-          }
-        });
 
-        return origReqHandler.call(this, req, res, parsedUrl);
+              // we'll collect this data in a more targeted way in the event processor we added above,
+              // `addRequestDataToEvent`
+              delete transaction.metadata.request;
+
+              // Push `transaction.finish` to the next event loop so open spans have a chance to finish before the
+              // transaction closes
+              setImmediate(() => {
+                transaction.finish();
+              });
+            }
+          });
+        }
       }
+
+      return origReqHandler.call(this, req, res, parsedUrl);
     });
   };
 
   return wrappedReqHandler;
 }
 
+/**
+ * Wrap the given method in order to use the parameterized path passed to it in the transaction name.
+ *
+ * @param origMethod Either `ensureApiPage` (called for every API request) or `findPageComponents` (called for every
+ * page request), both from the `Server` class
+ * @returns A wrapped version of the given method
+ */
+function makeWrappedMethodForGettingParameterizedPath(
+  origMethod: ApiPageEnsurer | PageComponentFinder,
+): WrappedApiPageEnsurer | WrappedPageComponentFinder {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const wrappedMethod = async function(this: Server, parameterizedPath: string, ...args: any[]): Promise<any> {
+    const transaction = getActiveTransaction();
+
+    // replace specific URL with parameterized version
+    if (transaction && transaction.metadata.request) {
+      // strip query string, if any
+      const origPath = transaction.metadata.request.url.split('?')[0];
+      transaction.name = transaction.name.replace(origPath, parameterizedPath);
```

Comment:
Why do we do a .replace here instead of just overwriting the URL?

Reply:
Also will this affect users manual transactions as well?

Reply:
I used `replace` because some of the transaction names have the method, and I only want to overwrite the URL part. As for manual transactions, it would only affect them if they a) store the correct metadata on the transaction, and b) have a request URL as part or all of their transaction name. Though not _technically_ impossible... it still feels unlikely.